### PR TITLE
Replication

### PR DESCRIPTION
**Pull request by newpoo**
This PR implement two new modes for replication.
Mode 1: ack client after committing to the master and write the update to the tcp connection to at least one of the slaves.
Mode 2: ack client after committing to the master and at least one of the slaves.

Besides the new unittests, I also manually tested it on the test cluster.

### CODE REVIEW

**Comment by newpoo** [comment]
kick build

File: rocksdb_replicator/max_number_box.h [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    , waiters_() {}
+
+  ~MaxNumberBox() {
+    // it's client's responsibility to ensure there is no pending calls when the
+    // destructor is called.
+    CHECK(waiters_.empty());
+  }
+
+  // no copy or move
+  MaxNumberBox(const MaxNumberBox&) = delete;
```
Comment (shuzhang1989) [inline rocksdb_replicator/max_number_box.h]: if it is move constructor, should it be MaxNumberBox&&

File: rocksdb_replicator/max_number_box.cpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+#include "folly/Likely.h"
+
+namespace replicator { namespace detail {
+
+void MaxNumberBox::post(const uint64_t num) {
+  std::vector<Waiter*> waiters_to_notify;
+  std::vector<Waiter*> waiters_to_wait;
+
+  {
+    std::unique_lock<std::mutex> lk(mtx_);
```
Comment (shuzhang1989) [inline rocksdb_replicator/max_number_box.cpp]: in highqps condition, will mutex be less performant than other lock implementations?

File: rocksdb_replicator/max_number_box.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  // still holding mtx_.
+  // This could make a pending thread return true from wait() or wait_for() due
+  // to spurious wakeup or timeout. This is benign. Because we use the pred versions
+  // of wait() and wait_for(). Also concurrent calling notify_one() is safe.
+  std::for_each(waiters_to_notify.begin(), waiters_to_notify.end(),
+                [] (Waiter* w) { w->cv.notify_one(); });
+}
+
+bool MaxNumberBox::wait(const uint64_t num, const uint64_t timeout_ms) {
+  thread_local Waiter me;
```
Comment (shuzhang1989) [inline rocksdb_replicator/max_number_box.cpp]: add some comment on the reason of thread local?

File: rocksdb_replicator/max_number_box.h (hunk 1 above)
Comment (newpoo) [inline rocksdb_replicator/max_number_box.h]: move constructor is implicilty deleted if the copy constructor is explicitly deleted.

File: rocksdb_replicator/max_number_box.cpp (hunk 2 above)
Comment (newpoo) [inline rocksdb_replicator/max_number_box.cpp]: Will explain it a little in the comment

File: rocksdb_replicator/max_number_box.cpp (hunk 3 above)
Comment (newpoo) [inline rocksdb_replicator/max_number_box.cpp]: Will explain it a little in the comment

**Pull request merged**